bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/handler/OpenThermGatewayHandler.java

```diff
@@ -0,0 +1,273 @@
+                        @Nullable
+                        Unit<?> unit = dataItem.getUnit();
+                        state = (unit == null) ? new DecimalType(value) : new QuantityType<>(value, unit);
+                        break;
+                    case DOWTOD:
+                        break;
+                }
+
+                if (state != null) {
+                    logger.debug("Received update for channel '{}': {}", channelId, state.toFullString());
+                    updateState(channelId, state);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void handleRemoval() {
+        logger.debug("Removing OpenTherm Gateway handler");
+        disconnect();
+        super.handleRemoval();
+    }
+
+    @Override
+    public void dispose() {
+        disconnect();
+        super.dispose();
+    }
+
+    private boolean checkConnection() {
+        @Nullable
+        OpenThermGatewayConnector conn = connector;
+
+        if (conn != null && conn.isConnected()) {
+            return true;
+        }
+
+        return connect();
+    }
+
+    private boolean connect() {
+        @Nullable
+        OpenThermGatewayConfiguration conf = config;
+
+        disconnect();
+
+        if (conf != null) {
+            logger.debug("Starting OpenTherm Gateway connector");
+
+            explicitDisconnect = false;
+
+            connector = new OpenThermGatewaySocketConnector(this, conf.ipaddress, conf.port);
+
+            Thread thread = new Thread(connector, "OpenTherm Gateway Binding - socket listener thread");
+            thread.setDaemon(true);
+            thread.start();
+
+            logger.debug("OpenTherm Gateway connector started");
+
+            return true;
+        }
+
+        return false;
+    }
+
+    private void disconnect() {
+        @Nullable
+        OpenThermGatewayConnector conn = connector;
+
+        if (conn != null) {
+            if (conn.isConnected()) {
+                logger.debug("Stopping OpenTherm Gateway connector");
+
+                explicitDisconnect = true;
+                conn.stop();
+            }
+
+            connector = null;
+        }
+    }
+
+    private String getGatewayCodeFromChannel(String channel) throws IllegalArgumentException {
+        switch (channel) {
+            case OpenThermGatewayBindingConstants.CHANNEL_OVERRIDE_SETPOINT_TEMPORARY:
+                return GatewayCommandCode.TemperatureTemporary;
+            case OpenThermGatewayBindingConstants.CHANNEL_OVERRIDE_SETPOINT_CONSTANT:
+                return GatewayCommandCode.TemperatureConstant;
+            case OpenThermGatewayBindingConstants.CHANNEL_OUTSIDE_TEMPERATURE:
+                return GatewayCommandCode.TemperatureOutside;
+            case OpenThermGatewayBindingConstants.CHANNEL_OVERRIDE_DHW_SETPOINT:
+                return GatewayCommandCode.SetpointWater;
+            case OpenThermGatewayBindingConstants.CHANNEL_SEND_COMMAND:
+                return "";
```

Comment:
Is `""` a valid gateway code? If not maybe you should allow this method to return null instead.

Reply:
Whether it's 'valid' is up to the code I guess.. it is possible to send raw commands instead of predefined commands. So either you send a predefined command based on a channel, in that case the channel is mapped to a particular command. Or you can send a raw command through the SendCommand channel, in which case you have to provide the command yourself.

The easiest way was to simple return "" so that it get's processed or concatenated as an empty string, followed by whatever the user had entered. The combination should produce a valid command, but that's up to the user. I changed it to null though, and added a check on the calling code.